app/models/reports/repository.rb

```diff
@@ -1,64 +1,143 @@
 module Reports
   class Repository
+    include Memery
+    PERCENTAGE_PRECISION = 0
+
     def initialize(regions, periods:, with_exclusions: false)
       @regions = Array(regions)
-      @regions_by_id = @regions.group_by { |r| r.id }
       @with_exclusions = with_exclusions
 
       @periods = if periods.is_a?(Period)
-        Range.new(periods, periods).to_a
+        Range.new(periods, periods)
       else
-        periods.to_a
+        periods
       end
     end
 
-    attr_reader :regions, :periods
-    attr_reader :regions_by_id
+    attr_reader :periods
+    attr_reader :regions
     attr_reader :with_exclusions
 
     delegate :cache, :logger, to: Rails
 
-    def controlled_patients_count
-      cached_query(:controlled_patients_count) do |item|
-        control_rate_query.controlled(item.region, item.period, with_exclusions: with_exclusions).count
+    # Uses Memery to memoize a method, but takes into account our force_cache setting. If force_cache is true,
+    # a caller is asking for all values to be retrieved fresh from the database, so we want to skip memoization.
```

Comment:
After our conversation on slack, I think it might be worth adding some additional comments like:
```
force_cache is only used when warming the cache and do not want to memoize in that case
```

Reply:
you may be interested in https://github.com/simpledotorg/simple-server/pull/2146